### ISSUE TITLE
MODS with BSAs containing animation files loaded higher then FNIS cause it to error.

### ISSUE THREAD

**Issue by GrantSP** [TanninOne/modorganizer#654]
This is an offshoot issue found during the the discussion about FNIS and Vilja in #647 .

I have noted that when the mod in question, Vilja but now also observable with USLEEP's BSA, is placed in a position HIGHER than the FNIS mod, running FNIS errors out with a message saying it can't find the specified animation file: _0_master.hkx_.

If either of those mods are loaded in MO's left-hand pane BEFORE FNIS, ie. with a lower priority than the FNIS mod, the tool runs correctly and completes its operation.

Here are two logs, one with a BSA before and one after.
[Before or with a LOW priority.](http://pastebin.com/f6QAwHvS)
[After or with a HIGH priority.](http://pastebin.com/XJzN3Gzy)

However, as the number of animation mods that contain animation files inside BSAs is very low, possibly Vilja being the only example and USLEEP will never be installed that low in the order this can be considered a very low priority case.
Add to that BSA management will be dropped in MO 2 and it may even be thought of as a non-issue.

However as a matter of courtesy I thought you should see the logs to support the claims made elsewhere about this minor issue.

**Comment by TanninOne** [comment]
I've tried to replicate your setup as much as possible but I still can't replicate the issue. Could you please verify that removing the wrye bashed patch doesn't change anything (it really shouldn't)?

Are you running MO as administrator?
Do you use a language other than english inside fnis?
What patches are you using in the test?
Could you go to the data tab, and from the right-click menu click "Write to file", and then send me that file? (maybe pack it as it's probably around 1MB)
Also, please attach the file "tools\GenerateFNIS_for_Users\FNISmodList.txt" and DebugData.txt?. For some reason, on your system there is an exception while processing one of those files (probably).

**Comment by GrantSP** [comment]
Setup is VERY bare. Only vanilla Skyrim with FNIS and Vilja. (Though I can replicate this error by substituting the Vilja mod's BSA with USLEEP's BSA.)
Skyrim is on **D:\Steam\steamapps\common\Skyrim\**
Mod Organizer is on: **D:\Games\MO\TES5\**

Only mods/tools installed are FNIS and Vilja. This was to examine the initial issue report.
There is no bashed patch.

Am I running a as administrator?  --- No. All my utilities are installed into non-UAC controlled folders.
Non-English  ---  No. Running as English
Data tab --- [MO-Data-tab.txt](https://github.com/TanninOne/modorganizer/files/223411/MO-Data-tab.txt)

FNIS doesn't get to that stage of running so those files are not generated.

**Comment by TanninOne** [comment]
Thanks, I'm now able to reproduce this, and I now know the reason. This will happen if the highest-priority mod that contains a 0_master.hkx has this file only in a bsa but there is another, lower-priority mod that has this file as a loose file.

tl;dr; explanation is that this is caused by an ancient hack I have in MO for performance reasons. Fixing this would be the clean thing to do but quite a bit of work and decrease performance for all applications, including the game itself. This problem will not occur with MO2. And it apparently affects only fnis.

longer explanation: As I explained in the other thread, if the highest priority variant of a file is in a bsa, MO on initialization doesn't add that file to the vfs, so that the game uses the one from the bsa and not the loose one. This is the basis for MOs bsa management.
Now unfortunately, when you retrieve a directory listing, MO can't provide all the info the OS would give for a listing from the vfs, so instead, MO creates a merged listing from all mod directories, and this does include even those files not added to the vfs.

So you might wonder why this is not a problem with other tools. As a rule, when coding you can't rely on the fs to remain the same between two calls. At any time an outside application could be deleting files. Also, you may find that on accessing a file, even though the file is there, you can't access it because of permissions or because another application has exclusive access.
This is why software will usually just try to do with a file whatever it needs and, in case of an error, deal with that error accordingly (i.e. "oh, deleting the file failed because the file is already gone, that's fine then").
But not fnis: it appears to retrieve a directory listing and continues to work under the assumption that the listing remains valid unless fnis itself changes it and it will error out if this assumption doesn't hold true.

To rephrase: MO has a performance optimization that works under the assumption that other tools won't work under false assumptions...

**Comment by GrantSP** [comment]
Excellent! I knew that was the reason... not.
Glad you were able to track it down so soon, now back to the grindstone for you and get MO 2 up and ready! ;)

**Comment by REALfore** [comment]
All right. lol

> tl;dr; explanation is that this is caused by an ancient hack I have in MO for performance reasons. Fixing this would be the clean thing to do but quite a bit of work and decrease performance for all applications, including the game itself. 

> Now unfortunately, when you retrieve a directory listing, MO can't provide all the info the OS would give for a listing from the vfs, so instead, MO creates a merged listing from all mod directories, and this does include even those files not added to the vfs.

but still

> MO has a performance optimization that works under the assumption that other tools (FNIS) won't work under false assumptions...

lol

- Where did FNIS have any problem with that used manually or under NMM?

- Where was is communicated that MO needs special consideration due to hacking and performance optimization? And why should FNIS care about performance optimizations after all?

- And most of all: do you still want to ignore that most of this delayed execution was only implemented as a work-around for MO's most nagging FNIS related bug. Its inability to deal with parallel processes (hkxcmd).

Ok, now ban me as annonced in the other thread. That will help you not to have to read the truth. And help me not to feel obliged to respond to your pathetic blames.

**Comment by TanninOne** [comment]
Did you have a stroke or something? I'm really starting to get concerned; something's not quite right with you...
Anyway, I won't watch you spam another issue with your pointless hostility.

**Comment by GrantSP** [comment]
I think this also can be considered closed.

**Comment by Kevdrag** [comment]
@REALfore troll XD

@GrantSP Y la moraleja es?... I mean, in conclution... what would be the solution then? If it's a FNIS problem and MO can't workaround it what should users do? Deleting the loose files like Emma (Vilja autor) said in the nexus page?

**Comment by GrantSP** [comment]
Yes.
